aptos-move/aptos-aggregator/src/aggregator_extension.rs

```diff
@@ -492,8 +486,117 @@ impl AggregatorData {
+                    .get_aggregator_v2_value(&id, AggregatorReadMode::Aggregated)
+                    .map_err(|e| {
+                        extension_error(format!(
+                            "Could not find the value of the aggregator: {}",
+                            e
+                        ))
+                    })?;
+                entry.insert(AggregatorSnapshot {
+                    id,
+                    state: AggregatorSnapshotState::Reference {
+                        speculative_value: SnapshotValue::try_from(value_from_storage)?,
+                    },
+                })
+            },
+            Entry::Occupied(entry) => entry.into_mut(),
+        };
+        Ok(snapshot)
+    }
+
+    pub fn read_snapshot(
+        &mut self,
+        id: AggregatorID,
+        resolver: &dyn AggregatorResolver,
+    ) -> PartialVMResult<SnapshotValue> {
+        // need to clone here, so we can call self.read_snapshot below.
+        let snapshot_state = self.get_snapshot(id, resolver)?.state.clone();
+        match snapshot_state {
+            AggregatorSnapshotState::Create { value } => Ok(value),
+            AggregatorSnapshotState::Delta {
+                base_aggregator,
+                delta,
+            } => match self
+                .aggregators
+                .get_mut(&AggregatorVersionedID::V2(base_aggregator))
+            {
+                Some(aggregator) => {
+                    let value = aggregator.read_aggregated_aggregator_value(resolver)?;
+                    Ok(SnapshotValue::Integer(expect_ok(
+                        BoundedMath::new(aggregator.max_value).unsigned_add_delta(value, &delta),
+                    )?))
+                },
+                None => resolver
+                    .get_aggregator_v2_value(&id, AggregatorReadMode::Aggregated)
+                    .map_err(|e| {
+                        extension_error(format!(
+                            "Could not find the value of the aggregator: {}",
+                            e
+                        ))
+                    })
+                    .and_then(SnapshotValue::try_from),
+            },
+            AggregatorSnapshotState::Derived {
+                base_snapshot,
+                formula,
+            } => {
+                let base = self.read_snapshot(base_snapshot, resolver)?;
+                match base {
+                    SnapshotValue::Integer(v) => Ok(SnapshotValue::String(formula.apply(v))),
+                    SnapshotValue::String(_) => Err(code_invariant_error(
+                        "Tried calling concat on String SnapshotValue",
+                    )),
+                }
+            },
+            AggregatorSnapshotState::Reference { speculative_value } => {
+                Ok(speculative_value.clone())
+            },
+        }
+    }
+
+    pub fn string_concat(
+        &mut self,
+        id: AggregatorID,
+        prefix: Vec<u8>,
+        suffix: Vec<u8>,
+    ) -> AggregatorID {
```

Comment:
By the name string_concat, ideally I would assume it concatenates two or more strings and returns the concatenated string. Here it creates a string snapshot and returns a string snapshot. Should we change the name of the function?

Reply:
it's too late, move code is in 1.7

Reply:
@vusirikala note that it's a method on AggregatorData, not a general string_concat